cornac 2.3.3

ifitchet opened the pull request:
cornac 2.3.3 

**Destination channel:** Defaults

### Links

- [PKG-8151]
- dev_url:        https://github.com/PreferredAI/cornac/tree/v2.3.3
- conda_forge:    https://github.com/conda-forge/cornac-feedstock
- pypi:           https://pypi.org/project/cornac/2.3.3
- pypi inspector: https://inspector.pypi.io/project/cornac/2.3.3

### Explanation of changes:

- new version number
- build `py>=39` noting:
  - we can't use the original 1.10.* pinnings as `numpy {{ numpy }}` from `cbc.yaml` implies 2.x
  - `scipy 1.13` which doesn't support `py313`
  - `scipy 1.14+` which doesn't support `py39`


[PKG-8151]: https://anaconda.atlassian.net/browse/PKG-8151?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ifitchet committed:
initial updates

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_323849_bzm/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:86: invalid_url: https://cornac.readthedocs.io : Not reachable: 403
===== Final Report: =====
0 Errors and 1 Warning were found

</details>

ifitchet committed:
remove staging channel

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_b6q_kj20s1/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:86: invalid_url: https://cornac.readthedocs.io : Not reachable: 403
===== Final Report: =====
0 Errors and 1 Warning were found

</details>

Andrii-Osipov commented on recipe/meta.yaml:
```diff
@@ -1,51 +1,56 @@
-    sha256: ada6d09bd78bbd5b7cd65d2eb8599146e4b8f52ef44651cb25fa6a019d36aa56
+    sha256: 6253d932c7fd65126c3be948134e5d04ecb756f7bf79a081ec0e32dc0dd1df39
   # for upstream tests and their pytest.ini
   - url: https://github.com/PreferredAI/cornac/archive/refs/tags/v{{ version }}.tar.gz
```
```suggestion
  - url: https://github.com/PreferredAI/{{ name }}/archive/refs/tags/v{{ version }}.tar.gz
```

Andrii-Osipov replied:
Why just not use only GitHub and avoid pypi?

ifitchet replied:
Well, as you can see from your suggested change I just left the `source.url`s as is without reading them.  😄 

We should get more consistent on how we use upstreams -- I appreciate that with psubs we've tended to be all or nothing with GH tarballs.

Andrii-Osipov commented on recipe/meta.yaml:
```diff
@@ -1,51 +1,56 @@
-    - setuptools
-    - libgomp  # [linux]
+    - cython >=0.29.21
+    - numpy {{ numpy }}
```
https://github.com/PreferredAI/cornac/blob/v2.3.3/pyproject.toml#L6
numpy >2
in cbc.yaml: https://github.com/AnacondaRecipes/aggregate/blob/master/conda_build_config.yaml#L100
correct only for py313

ifitchet replied:
Oh yes, that is technically, true.  Good spot.  Although more technically, still, `cbc.yaml` is saying `2.0` as in `2.0.*` (and similarly for `2.1`).
I fancy that the upstream didn't really mean `>2.0.0` as if `2.0.0` _itself_ was broken, however they should (and do) get `2.0.2` and similarly `2.1.3` during build.  Both get `2.2.4` at runtime.

Andrii-Osipov commented on recipe/meta.yaml:
```diff
@@ -1,51 +1,56 @@
 build:
-  skip: True  # [py<38 or py>312]
+  # scipy 1.14 is py310+
+  skip: true  # [py<310]
```
https://github.com/PreferredAI/cornac/blob/v2.3.3/pyproject.toml#L22
requires-python = ">=3.9"
You can pin different scipy versions for different Python versions.

ifitchet replied:
OK, let me revisit that with one of those `[py==39]`-style selectors.  (Never used them before -- very exciting!)

Andrii-Osipov commented on recipe/meta.yaml:
```diff
@@ -1,51 +1,56 @@
+    # scipy.linalg.cython_blas, 1.1[34].* are OK but
+    # - 1.13 does not support py313
+    # - 1.14 does not support py39
+    - scipy 1.14.1
```
Where did you find that we must use this 1.14.1 version?

ifitchet replied:
Trial and error.
I'm not sure why `1.15.x` doesn't work.  Maybe it dropped some code.  🤷

ifitchet replied:
s/Trial and error/Science!/

We know that `numpy {{ numpy }}` is going to ask for a `2.x` `numpy` from `cbc.yaml` for which we need a compatible `scipy`.  Let's have a grovel at `py311` variants:

```bash
$conda search scipy=*=py311* -i | grep -E '(^scipy|numpy)'
scipy 1.9.3 py311hc76d9b0_1
  - numpy >=1.23,<1.26.0
...
  - numpy >=1.23.5,<1.29
scipy 1.13.1 py311hac8794a_1
  - numpy >=1.22.4,<2.3
scipy 1.14.1 py311hac8794a_0
  - numpy >=1.23.5,<2.3
scipy 1.15.1 py311hac8794a_0
  - numpy >=1.23.5,<2.5
...
```

From which you can see we need `scipy 1.13.1` (build 1) to satisfy the `numpy` constraint.  Ditto the other `py*` ones apart from `py313` which was only built in `scipy 1.14.1`.

Andrii-Osipov commented on recipe/meta.yaml:
```diff
@@ -1,51 +1,56 @@
+    # add further pinnings for numpy and scipy
+    - numpy >2.0.0
+    - scipy
+    - libgomp     # [linux]
```
Why do we need that lib?

Andrii-Osipov replied:
run_exports problem?

ifitchet replied:
Yes.  You'll get an over/under/sideways linking complaint (explicitly for `libgomp`) if you don't.
Perhaps I should include the error message if it's missing.

ifitchet replied:
actually:

```
  ERROR (cornac,lib/python3.9/site-packages/cornac/utils/fast_dot.cpython-39-aarch64-linux-gnu.so): Needed DSO lib/libgomp.so.1 found in ['defaults/linux-aarch64::_openmp_mutex==5.1=51_gnu']
```

So, uh, `python`?

ifitchet replied:
In practice you get an overdepending error against Cython `.so` files which want `libgomp`.

ifitchet committed:
address comments:

- build for py39 because...

- clarify which scipy versions (compatible with numpy 2.x from
  cbc.yaml) are required by py* versions

- don't actually need the C compiler

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_33q8wghew7/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:84: invalid_url: https://cornac.readthedocs.io : Not reachable: 403

===== ERRORS =====
- clone/recipe/meta.yaml:30: cython_needs_compiler: Cython generates C code, which will need to be compiled
===== Final Report: =====
1 Error and 1 Warning were found

</details>

ifitchet committed:
re-adding the C compiler to appease the linter

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_54om9vuj53/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:85: invalid_url: https://cornac.readthedocs.io : Not reachable: 403
===== Final Report: =====
0 Errors and 1 Warning were found

</details>

Andrii-Osipov commented on recipe/meta.yaml:
```diff
@@ -57,7 +61,7 @@ test:
     - pytest
     # - pytest-pep8 # unavailable
```
remove this line

ifitchet committed:
removed comment about unavailable package

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_bdya_2cty1/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:84: invalid_url: https://cornac.readthedocs.io : Not reachable: 403
===== Final Report: =====
0 Errors and 1 Warning were found

</details>

skupr-anaconda reviewed (approved):
LGTM

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_d32vgskzp5/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:84: invalid_url: https://cornac.readthedocs.io : Not reachable: 403
===== Final Report: =====
0 Errors and 1 Warning were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_8bc9dhn2us/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:84: invalid_url: https://cornac.readthedocs.io : Not reachable: 403
===== Final Report: =====
0 Errors and 1 Warning were found

</details>
